src/mlir/dialect/VeronaDialect.cc

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "VeronaDialect.h"
+
+#include "VeronaOps.h"
+#include "mlir/IR/DialectImplementation.h"
+#include "mlir/IR/StandardTypes.h"
+
+using namespace mlir;
+using namespace mlir::verona;
+
+//===----------------------------------------------------------------------===//
+// Verona dialect.
+//===----------------------------------------------------------------------===//
+
+VeronaDialect::VeronaDialect(mlir::MLIRContext* context)
+: Dialect(getDialectNamespace(), context)
+{
+  addOperations<
+#define GET_OP_LIST
+#include "dialect/VeronaOps.cpp.inc"
+    >();
+  addTypes<IntegerType>();
+  allowUnknownOperations();
+  allowUnknownTypes();
+}
+
+Type VeronaDialect::parseType(DialectAsmParser& parser) const
+{
+  StringRef keyword;
+  if (parser.parseKeyword(&keyword))
+    return Type();
+
+  if (keyword.startswith("U"))
+  {
+    size_t width = 0;
+    if (keyword.substr(1).getAsInteger(10, width))
+    {
+      parser.emitError(parser.getNameLoc(), "unknown verona type: ") << keyword;
+      return Type();
+    }
+    return IntegerType::get(getContext(), width, false);
+  }
+
+  parser.emitError(parser.getNameLoc(), "unknown verona type: ") << keyword;
+  return Type();
+}
+
+void VeronaDialect::printType(Type type, DialectAsmPrinter& os) const
+{
+  switch (type.getKind())
+  {
+    case VeronaTypes::Integer:
+    {
+      auto iTy = type.cast<IntegerType>();
+      if (iTy.getSign())
+      {
+        os << "S";
+      }
+      else
+      {
+        os << "U";
+      }
+      os << iTy.getWidth();
+      return;
+    }
+    default:
+      llvm_unreachable("unexpected 'verona' type kind");
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// Verona types.
+//===----------------------------------------------------------------------===//
+
+namespace mlir::verona::detail
+{
+  struct IntegerTypeStorage : public ::mlir::TypeStorage
+  {
+    size_t width;
+    enum SignType
+    {
+      Unknown,
+      Unsigned,
+      Signed
+    };
+    unsigned sign;
```

Comment:
Because `std::get<1>(key)` returns unsigned, which is defined as `KeyTy` that way so `TypeStorage` can use it internally for a `DenseMap` key, given that `SignType` is a local type.

Reply:
I don't think I understand.  Why can't `KeyType` be `std::tuple<size_t, SignType>`?

Reply:
AFAICS, it's because that's used as a key to `DenseMap` and some of its own functionality isn't specialised for the local type. I haven't dug deep to make it work, but seeing the same pattern in other examples of MLIR, I just repeated the pattern and spent my time elsewhere.

We could perhaps revisit this issue once we have an actual type to implement.